flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/RecordSerializer.java

```diff
@@ -66,29 +66,33 @@ public boolean isFullBuffer() {
 	}
 
 	/**
-	 * Starts serializing and copying the given record to the target buffer
-	 * (if available).
+	 * Starts serializing the given record to an intermediate data buffer.
 	 *
 	 * @param record the record to serialize
-	 * @return how much information was written to the target buffer and
-	 *         whether this buffer is full
 	 */
-	SerializationResult addRecord(T record) throws IOException;
+	void serializeRecord(T record) throws IOException;
```

Comment:
The previous `RecordSerializer` also confuses me a lot and I have the same experience with you, because the previous `addRecord` and `continueWritingWithNextBufferBuilder`  methods can be called in arbitrary sequence and both returned `SerializationResult`.

In my current reconstruction, the method `serializeRecord` must be called firstly, and then the method `copyToBufferBuilder` is called to return the final `SerializationResult`. I think it seems a bit clearer than before.

I agree your above idea is good for separating these two methods further. But the `RecordSerializer` and `SerializedRecord` may be still close with each other. I think there are two ways to realize `SerializedRecord#copyToBufferBuilder`:
1. 
```
   public SerializedRecord(ByteBuffer lengthBuffer, ByteBuffer dataBuffer) {
	}

    CopingResult copyToBufferBuilder(BufferBuilder bufferBuilder) {
        // copy  lengthBuffer
       // copy dataBuffer
      //  get CopingResult
	}
```

So this way the `SerializedRecord` can only see `lengthBuffer` and `dataBuffer`, and can not interact with `RecordSerializer`. Maybe we do not need do anything in `SerializedRecord#close()`.

2. 
  ```
 public SerializedRecord(RecordSerializer serializer) {
   }

    CopingResult copyToBufferBuilder(BufferBuilder bufferBuilder) {
        serializer.copyToBufferBuilder();
        //  get CopingResult
	}
```

This way the `SerializedRecord` can see and interact with `RecordSerializer`, but the only difference seems we separate the `SerializedRecord` and `CopyingResult`. And my current implementation is we hide the `SerializedRecord` and return `SerializationResult` which corresponds to `CopyingResult` as final result.

What do you think of the above ways?

Reply:
Re 1.

I was thinking about something along those lines, however I saw the need for `close()`, because currently `SpanningRecordSerializer` reuses for multiple records `byte[]` that hides in `dataBuffer = serializationBuffer.wrapAsByteBuffer();` call. Is there some other way to release/return this array back to `RecordSerializer`? Or maybe that proves that `RecordSerializer` and `SerializedRecord` should be indeed in one class?

Re 2. 

I'm not sure. This one seems like it would leak `RecordSerializer serializer` private fields to `SerializedRecord`.

All in all, I'm not sure if the improvement here is worth the effort and even if we would improve/simplify code here (especially this `close()` bothers me). Also I agree that your changes in `addRecord`, `continueWritingWithNextBufferBuilder` and `copyToBufferBuilder` makes this code actually easier all in all, so I would be fine with merging it as is (modulo my other comments) :)

Reply:
I agree with not exposing `RecordSerializer`'s internal private fields to `SerializedRecord` for safe concern.

Regarding with the way of reference of the `RecordSerializer` directly in `SerializedRecord`. The whole process in `RecordWriter` may be like this:

		SerializedRecord  serializedRecord = serializer.serializeRecord(record);
		for (int targetChannel = 0; targetChannel < numChannels; targetChannel++) {
                        //reset position for multiple copying
                        serializer.reset();
			BufferBuilder bufferBuilder = getBufferBuilder(targetChannel);
		        CopyingResult result = serializedRecord.copyToBufferBuilder(bufferBuilder);
		}
		serializer.prune();

The `SerializedRecord` only provides the method of `copyToBufferBuilder` and wraps the internal class `CopyingResult`, but the actual copying operation is still done in `RecordSerializer`. So the `RecordSerializer` also needs provide the method `copyToBufferBuilder`. From this point, we actually do not separate the two steps of `serialization` and `copy` from `RecordSerializer`. So I think maybe it is not very worth doing that. :)

Reply:
So maybe keep it as it is. Maybe in the future someone will be struck by a better idea.